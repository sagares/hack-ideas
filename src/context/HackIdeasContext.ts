import { createContext } from "react";
import { IdeasArray } from "../HackIdea.types";

type Employee = {
  empId: string;
};

type HackIdeasContextType = {
  currentUser: Employee;
  logout: () => void;
  setCurrentUser: (employee: Employee) => void;
  ideas: IdeasArray;
  setIdeas: (ideas: IdeasArray) => void;
};

const HackIdeasContext = createContext<Partial<HackIdeasContextType>>({});

export { HackIdeasContext };
