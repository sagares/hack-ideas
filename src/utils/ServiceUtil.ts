// Import the functions you need from the SDKs you need
import { initializeApp } from "firebase/app";
import { getDatabase, ref, onValue, set } from "firebase/database";
import {  Idea } from "../HackIdea.types";

const firebaseConfig = {
  apiKey: "AIzaSyC0RFn0RNnX8EoPHpAu-lf4EvN08aPKMCg",
  authDomain: "hack-ideas-c7cd7.firebaseapp.com",
  databaseURL: "https://hack-ideas-c7cd7-default-rtdb.firebaseio.com",
  projectId: "hack-ideas-c7cd7",
  storageBucket: "hack-ideas-c7cd7.appspot.com",
  messagingSenderId: "197788164361",
  appId: "1:197788164361:web:36a94bf614b4b5c830f3d8",
  measurementId: "G-660ZLLJJEB",
};

// Initialize Firebase
const app = initializeApp(firebaseConfig);
const database = getDatabase();

const registerEmployee = (empId: string) => {
  set(ref(database, "employees/" + empId), {
    empId,
  });
};

const getEmployeeById = (empId: string, callback: Function) => {
  let empRef = ref(database, "employees/" + empId);
  return onValue(empRef, (snapshot) => {
    callback(snapshot.val());
  });
};

const getEmployeeData = (callback: Function) => {
  let empRef = ref(database, "employees/");
  return onValue(empRef, (snapshot) => {
    callback(snapshot.val());
  });
};

const getIdeas = (callback: Function) => {
  let empRef = ref(database, "ideas/");
  return onValue(empRef, (snapshot) => {
    callback(snapshot.val());
  });
};

const registerIdea = (idea: Idea) => {
  const id = toCamelCase(idea.name);
  set(ref(database, "ideas/" + id), {
    ...idea,
    id,
    upvotes: [],
  });
};

const updateIdea = (ideaId: string, updatedIdea: Idea) => {
  set(ref(database, "ideas/" + ideaId), updatedIdea);
};

const toCamelCase = (sentenceCase: string) => {
  var out = "";
  sentenceCase.split(" ").forEach((el, idx) => {
    var add = el.toLowerCase();
    out += idx === 0 ? add : add[0].toUpperCase() + add.slice(1);
  });
  return out;
};

export {
  getEmployeeById,
  getEmployeeData,
  getIdeas,
  registerEmployee,
  registerIdea,
  updateIdea,
};
