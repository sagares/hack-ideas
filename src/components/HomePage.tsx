import React, { useContext, useEffect, useState } from "react";
import { getIdeas } from "../utils/ServiceUtil";
import HackIdeasContext from "../context";
import { Idea } from "../HackIdea.types";
import { useSearchableIdeas } from "../hooks";
import NavBar from "./nav-bar";
import RegisterIdea from "./register-idea";
import { Card, Filter } from ".";

import "./home-page.less";

const HomePage = () => {
  const { ideas, setIdeas } = useContext(HackIdeasContext);

  const [searchString, setSearchString] = useState("");
  const [orderAsc, serOrderAsc] = useState(true);
  const [sortBy, setSortBy] = useState("");
  const filteredIdeas = useSearchableIdeas(
    ideas,
    searchString,
    sortBy,
    orderAsc
  );

  const handleSortChange = (newSortBy: string) => {
    setSortBy((prevSortBy) => {
      serOrderAsc(prevSortBy === newSortBy ? !orderAsc : orderAsc);
      return newSortBy;
    });
  };

  useEffect(() => {
    getIdeas(setIdeas);
  }, []);

  return (
    <div className="hack-ideas--home-page">
      <NavBar />
      <Filter
        onFilter={setSearchString}
        searchText={searchString}
        sortBy={sortBy}
        orderAsc={orderAsc}
        onSort={handleSortChange}
      />
      <div className="hack-ideas--cards">
        {filteredIdeas?.length > 0 ? (
          <div className="hack-idea--grid">
            {filteredIdeas.map((idea: Idea) => {
              return <Card id={idea.id} key={idea.id} />;
            })}
          </div>
        ) : (
          <div className="card-panel red lighten-4 center-align">
            No ideas to show.
          </div>
        )}
      </div>
      <RegisterIdea />
    </div>
  );
};

export default HomePage;
