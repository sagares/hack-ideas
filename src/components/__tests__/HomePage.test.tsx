/**
 * @jest-environment jsdom
 */
import React from "react";
import { act, render } from "@testing-library/react";
import { getIdeas } from "../../utils/ServiceUtil";
import HackIdeasContext from "../../context";
import HomePage from "../HomePage";

const callbackMap = {};
const mockFilterComponent = jest.fn();
jest.mock("../../utils/ServiceUtil");
jest.mock("../filter/Filter", () => (props) => {
  for(let prop in props) {
    if(typeof props[prop] === "function") {
      callbackMap[prop] = props[prop];
    }
  };
  mockFilterComponent(props);
  return <div>Filter</div>;
});

describe("Home Page Component", () => {
  let element, component, setIdeas;

  beforeEach(() => {
    (getIdeas as jest.Mock).mockImplementation((cb) => {
      cb({ idea1: { name: "idea 1" } });
      return jest.fn();
    });

    setIdeas = jest.fn().mockImplementation(() => {});
    element = document.createElement("div");

    component = render(
      <HackIdeasContext.Provider
        value={{
          currentUser: { empId: "123456" },
          ideas: {
            idea1: {
              name: "idea 1",
              description: "idea 1",
              id: "idea1",
              tags: [],
              submittedBy: "123456",
              submittedOn: 100001,
            },
          },
          setIdeas,
        }}
      >
        <HomePage />
      </HackIdeasContext.Provider>,
      {
        container: document.body.appendChild(element),
      }
    );
  });

  afterEach(() => {
    document.body.removeChild(element);
  });

  test("renders hack ideas grid", () => {
    expect(component.container).toMatchSnapshot();
    expect(component.container.querySelector(".hack-idea--grid")).toBeDefined();
    expect(mockFilterComponent).toHaveBeenCalledWith(
      expect.objectContaining({
        orderAsc: true,
        searchText: "",
        sortBy: "",
      })
    );
  });

  test("handles on sort callback", () => {
    act(() => {
      callbackMap["onSort"]("mock-sort-by");
    });
    expect(mockFilterComponent).toHaveBeenCalledWith(
      expect.objectContaining({
        orderAsc: true,
        searchText: "",
        sortBy: "mock-sort-by",
      })
    );
  });

  test("reverts ascending order if sort by is same as previous", () => {
    act(() => {
      callbackMap["onSort"]("new-sort-by");
    });
    act(() => {
      callbackMap["onSort"]("new-sort-by");
    });
    expect(mockFilterComponent).toHaveBeenCalledWith(
      expect.objectContaining({
        orderAsc: false,
        searchText: "",
        sortBy: "new-sort-by",
      })
    );
  });

  test("does not render ideas grid if ideas not defined", () => {
    component = render(
      <HackIdeasContext.Provider
        value={{
          currentUser: { empId: "123456" },
          ideas: null,
          setIdeas,
        }}
      >
        <HomePage />
      </HackIdeasContext.Provider>,
      {
        container: document.body.appendChild(element),
      }
    );

    expect(component.container).toMatchSnapshot();
    expect(component.container.querySelector(".hack-idea--grid")).toBe(null);
    expect(component.container.querySelector(".card-panel.red")).toBeDefined();
  });
});
