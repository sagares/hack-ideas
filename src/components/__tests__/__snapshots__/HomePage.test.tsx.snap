// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page Component does not render ideas grid if ideas not defined 1`] = `
<div>
  <div
    class="hack-ideas--home-page"
  >
    <nav>
      <div
        class="nav-wrapper"
      >
        <a
          class="brand-logo"
        >
          { Hack Ideas }
        </a>
        <ul
          class="right hide-on-med-and-down"
        >
          <li>
            <a
              class="logout-btn"
            >
              Logout
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <div
      class="hack-ideas--filter row"
    >
      <div
        class="search-bar col s12 m12 l4"
      >
        <nav>
          <div
            class="nav-wrapper"
          >
            <form>
              <div
                class="input-field grey lighten-4"
              >
                <input
                  id="search"
                  type="search"
                  value=""
                />
                <label
                  class="label-icon"
                  for="search"
                >
                  <i
                    class="material-icons"
                  >
                    search
                  </i>
                </label>
                <i
                  class="clear-icon material-icons"
                >
                  close
                </i>
              </div>
            </form>
          </div>
        </nav>
      </div>
      <div
        class="filter-dropdown col s4 m2 l2"
      >
        <a
          class="dropdown-trigger btn-floating btn-large"
          data-target="dropdown1"
        >
          <i
            class="large material-icons"
          >
            sort
          </i>
        </a>
        <ul
          class="dropdown-content"
          id="dropdown1"
          tabindex="0"
        >
          <li
            class=""
            tabindex="0"
          >
            <a>
              <i
                class="material-icons"
              >
                access_time
              </i>
              By Created On
            </a>
          </li>
          <li
            class=""
            tabindex="0"
          >
            <a>
              <i
                class="material-icons"
              >
                thumb_up
              </i>
              By Upvotes
            </a>
          </li>
          <li
            class="divider"
            tabindex="-1"
          />
          <li
            tabindex="0"
          >
            <a>
              <i
                class="material-icons"
              >
                cached
              </i>
              Reset
            </a>
          </li>
        </ul>
      </div>
    </div>
    <div
      class="hack-ideas--cards"
    >
      <div
        class="card-panel red lighten-4 center-align"
      >
        No ideas to show.
      </div>
    </div>
    <div
      class="fixed-action-btn register-idea-btn"
    >
      <a
        class="btn-floating btn-large"
      >
        <i
          class="large material-icons"
        >
          add
        </i>
      </a>
    </div>
  </div>
</div>
`;

exports[`Home Page Component renders hack ideas grid 1`] = `
<div>
  <div
    class="hack-ideas--home-page"
  >
    <nav>
      <div
        class="nav-wrapper"
      >
        <a
          class="brand-logo"
        >
          { Hack Ideas }
        </a>
        <ul
          class="right hide-on-med-and-down"
        >
          <li>
            <a
              class="logout-btn"
            >
              Logout
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <div
      class="hack-ideas--filter row"
    >
      <div
        class="search-bar col s12 m12 l4"
      >
        <nav>
          <div
            class="nav-wrapper"
          >
            <form>
              <div
                class="input-field grey lighten-4"
              >
                <input
                  id="search"
                  type="search"
                  value=""
                />
                <label
                  class="label-icon"
                  for="search"
                >
                  <i
                    class="material-icons"
                  >
                    search
                  </i>
                </label>
                <i
                  class="clear-icon material-icons"
                >
                  close
                </i>
              </div>
            </form>
          </div>
        </nav>
      </div>
      <div
        class="filter-dropdown col s4 m2 l2"
      >
        <a
          class="dropdown-trigger btn-floating btn-large"
          data-target="dropdown1"
        >
          <i
            class="large material-icons"
          >
            sort
          </i>
        </a>
        <ul
          class="dropdown-content"
          id="dropdown1"
          tabindex="0"
        >
          <li
            class=""
            tabindex="0"
          >
            <a>
              <i
                class="material-icons"
              >
                access_time
              </i>
              By Created On
            </a>
          </li>
          <li
            class=""
            tabindex="0"
          >
            <a>
              <i
                class="material-icons"
              >
                thumb_up
              </i>
              By Upvotes
            </a>
          </li>
          <li
            class="divider"
            tabindex="-1"
          />
          <li
            tabindex="0"
          >
            <a>
              <i
                class="material-icons"
              >
                cached
              </i>
              Reset
            </a>
          </li>
        </ul>
      </div>
    </div>
    <div
      class="hack-ideas--cards"
    >
      <div
        class="hack-idea--grid"
      >
        <div
          class="hack-idea--card card"
        >
          <div
            class="card-content"
          >
            <span
              class="card-title"
            >
              idea 1
            </span>
            <p>
              idea 1
            </p>
            <div
              class="card-chips"
            />
          </div>
          <div
            class="card-action"
          >
            <a
              class="card-action--button  disabled"
            >
              <i
                class="material-icons"
              >
                thumb_up
              </i>
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      class="fixed-action-btn register-idea-btn"
    >
      <a
        class="btn-floating btn-large"
      >
        <i
          class="large material-icons"
        >
          add
        </i>
      </a>
    </div>
  </div>
</div>
`;
